### PR TITLE
[DO NOT MERGE - Pending Upstream PyTorch PR 194033 merge] clip_grad_norm_: compute the total norm in float32 so it does not depend on the parallelism partition

### PR DESCRIPTION
**Pull request by QIU023**
**Summary**

`get_total_norm` returns the norm in the input tensors' dtype, so with `training.dtype="bfloat16"` both call sites compute the per-tensor norms and the norm-of-norms in bf16 (3-4 significant digits). The rounding happens per group, so under PP/EP the total depends on how params are grouped across ranks: two runs with bit-identical gradients but different pipeline splits report different `grad_norm` and clip differently. The default `training.dtype="float32"` is unaffected.

**Evidence**

At `f4e78188e`, 4 GPUs, both dtypes, patch on/off:

```
torchrun --nproc_per_node=4 -m torchtitan.train --module llama3 --config llama3_debugmodel \
  --debug.seed 42 --debug.deterministic --metrics.log_freq 1 --training.steps 2 \
  --training.dtype bfloat16 --parallelism.data_parallel_shard_degree 2 \
  --parallelism.pipeline_parallel_degree 2 --training.local-batch-size 2
```

| `--training.dtype` | patch | step 1 | step 2 |
|---|---|---|---|
| float32 | before / after | 1.4509 | 1.6336 |
| bfloat16 | before | 1.4453 | 1.6328 |
| bfloat16 | after | **1.4508** | 1.6343 |

float32 is byte-identical before/after; patched bf16 lands next to float32 (1.4508 vs 1.4509).

**Fix**

Run those two norm ops with `dtype=torch.float32` via a local `_get_total_norm_fp32` that otherwise mirrors `get_total_norm`; gradients and the clip multiply are unchanged. It lives in this file rather than `torch.nn.utils` because `get_total_norm` has no `dtype` parameter and adding one is a PyTorch API change — happy to propose that upstream, at which point this helper reduces to a one-line call.

### CODE REVIEW

File: torchtitan/distributed/utils.py [hunk 1]
```diff
@@ -595,6 +595,45 @@ def set_pg_timeouts(
         torch.distributed.set_timeout(timeout, group)
 
 
+@torch.no_grad()
+def _get_total_norm_fp32(
```
Comment (tianyu-l) [inline torchtitan/distributed/utils.py:599]: sorry I couldn't really understand the PR summary which seems to be written by AI.

Before vs. after this change, what ops that are used to performed in bf16 are not performed in fp32?

Comment (QIU023) [reply torchtitan/distributed/utils.py:599]: Hi Tianyu, sorry for above unclean PR summary, will make sure to send clear and simple conclusion for further fix PRs content during collaborating with AI for K3 support, I have also simplified above PR summary

the answer is:

The norm operator computation is structurally identical before and after — per-tensor norms, then each group's norm-of-norms, then the cross-mesh combine; 

The fix only changes the dtype those two norm ops accumulate in (`get_total_norm` keeps the input dtype, so with bf16 grads they ran in bf16). Each group's partial norm is rounded before the combine in both cases — at bf16's $2^{-8}$ step that rounding is large enough that different PP/EP layouts of the same gradients yield visibly different totals, at fp32's $2^{-23}$ it is negligible, so all layouts agree. The gradients themselves stay bf16.

Comment (tianyu-l) [reply torchtitan/distributed/utils.py:599]: I guess the problem is that `torch.nn.utils.get_total_norm` doesn't support configurable norm dtype, or default to the safer fp32. Is that correct?

@janeyx99 do you think we can improve this from pytorch core side?

Comment (QIU023) [reply torchtitan/distributed/utils.py:599]: Yes! on torchtitan side I think we can always use the safer `fp32` for gradient reduction across the parallelism mesh to minimize the rounding during mesh reduction for safety

once the pytorch core `torch.nn.utils.get_total_norm` supports input `dtype` arg, this PR mirror helper function can be deleted with a cleanup PR, LMK if you agree with this proposal and I can add some docstring in this PR for reminding cleanups waiting `torch.nn.utils.get_total_norm`

Comment (janeyx99) [reply torchtitan/distributed/utils.py:599]: @QIU023 I think it's worth opening an issue for it on the pytorch/pytorch side. I wish GH was working rn so I can see how you implemented the fp32 version, but it may need design given no precedent on ops taking in "dtype" argument. Some questions I would have for design:
- is it unambiguous if we have a parameter "accumulate_dtype" which operations are done in this dtype and not?
- does this require changing the apis of vector_norm and _foreach_norm? (based on your lines change, I'm thinking no?)

Comment (janeyx99) [reply torchtitan/distributed/utils.py:599]: Ah GH is loading now! Nice it looks like only the norm ops get the dtype and they already support taking in dtypes. I'm more confident this can be resolved in pytorch/pytorch if you open an issue/PR and tag me!

Comment (QIU023) [reply torchtitan/distributed/utils.py:599]: Sure Jane! Will raise the PR for the `dtype` arg adding for ` torch.nn.utils.get_total_norm` in pytorch/pytorch very soon! all verifications are done there and once that part is merged I will update this titan PR so @tianyu-l can review the new content diff with the duplicate helper function deleted after the next pytorch nightly release available for torchtitan

Comment (QIU023) [reply torchtitan/distributed/utils.py:599]: Hi Jane! @janeyx99  upstream pytorch PR opened: https://github.com/pytorch/pytorch/pull/194033, please check over there, thanks! this one I will add pending note so Tianyu can review it later
